src/other/broadcasting.jl

```diff
@@ -0,0 +1,94 @@
+struct DummyColumnValue end
+
+Base.broadcastable(df::AbstractDataFrame) = Matrix(df)
+
+Base.maybeview(df::AbstractDataFrame, idxs...) = view(df, idxs...)
+
+function Base.maybeview(df::AbstractDataFrame, idxs)
+    if idxs isa Symbol
+        if !haskey(df, idxs)
+            if !(df isa DataFrame)
+                throw(ArgumentError("column $idxs not found"))
+            end
+            df[idxs] = Vector{DummyColumnValue}(undef, nrow(df))
+            return view(df, [idxs])
+        end
+    end
+    if idxs isa AbstractVector{Symbol}
+        for idx in idxs
+            if !haskey(df, idx)
+                if !(df isa DataFrame)
+                    throw(ArgumentError("column $idxs not found"))
+                end
+                df[idx] = Vector{DummyColumnValue}(undef, nrow(df))
+            end
+        end
+    end
+    view(df, idxs)
+end
+
+function Base.copyto!(df::SubDataFrame, bc)
+    m = Base.materialize(bc)
+    if ncol(df) == 1 && eltype(df[1]) === DummyColumnValue
+        if maximum(ndims.(Base.Broadcast.flatten(bc).args)) > 1
+            throw(DimensionMismatch("cannot broadcast array to have fewer dimensions"))
+        end
+    end
+    for col in axes(df, 2)
+        if eltype(df[col]) === DummyColumnValue
+            colname = names(df)[col]
+            T = reduce(promote_type, typeof.(view(m, :, col)))
```

Comment:
It feels like you should be able to fuse the `typeof.()` here into the function we pass to `reduce`.  Perhaps:

```suggestion
            T = reduce(((::Union{T,Type{T}}, ::Union{S,Type{S}}) where {T,S})->promote_type(T, S), view(m, :, col); init=Union{})
```

Reply:
I do not understand this `Type{T}` part in the signature. I would assume that it is not needed as otherwise I will a wrong result if e.g. my vector I want to reduce over is `[Int64, Float64]` - I will get `T` to be `Float64` and it should be `DataType`.

Reply:
Yeah, I was trying to be a bit too clever.  The idea stands, though — it would be nice to do the typeof operation in the reduction function itself instead of broadcasting (and allocating an array) for `typeof.()`.  It'll just take a bit smarter of a function definition.  The tricky part is simply that the array you're reducing over holds values but the intermediate results are types.  So the first time you call the reduction function you get two values, and then every subsequent time you get a type and a value.

Reply:
Thanks - this is what I thought and rightly `typeof.` is too expensive here (I did not want to optimize out everything in the first pass). Probably `mapreduce` instead of `reduce` should work here.